internal/controller/bootstrap/controlplane_bootstrap_controller.go

```diff
@@ -0,0 +1,479 @@
+
+func (c *ControlPlaneController) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	log := log.FromContext(ctx).WithValues("K0sControllerConfig", req.NamespacedName)
+	log.Info("Reconciling K0sControllerConfig")
+
+	// Lookup the config object
+	config := &bootstrapv1.K0sControllerConfig{}
+	if err := c.Get(ctx, req.NamespacedName, config); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("K0sControllerConfig not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get config")
+		return ctrl.Result{}, err
+	}
+
+	// Look up the owner of this config if there is one
+	configOwner, err := bsutil.GetConfigOwner(ctx, c.Client, config)
+	if apierrors.IsNotFound(errors.Cause(err)) {
+		// Could not find the owner yet, this is not an error and will rereconcile when the owner gets set.
+		log.Info("Owner not found yet, waiting until it is set")
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		log.Error(err, "Failed to get owner")
+		return ctrl.Result{}, err
+	}
+	if configOwner == nil {
+		log.Info("Owner is nil, waiting until it is set")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("kind", configOwner.GetKind(), "version", configOwner.GetResourceVersion(), "name", configOwner.GetName())
+
+	// Lookup the cluster the config owner is associated with
+	cluster, err := capiutil.GetClusterByName(ctx, c.Client, configOwner.GetNamespace(), configOwner.ClusterName())
+	if err != nil {
+		if errors.Cause(err) == capiutil.ErrNoCluster {
+			log.Info(fmt.Sprintf("%s does not belong to a cluster yet, waiting until it's part of a cluster", configOwner.GetKind()))
+			return ctrl.Result{}, nil
+		}
+
+		if apierrors.IsNotFound(errors.Cause(err)) {
+			log.Info("Cluster does not exist yet, waiting until it is created")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Could not get cluster with metadata")
+		return ctrl.Result{}, err
+	}
+
+	if annotations.IsPaused(cluster, config) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	scope := &Scope{
+		ConfigOwner: configOwner,
+		Cluster:     cluster,
+	}
+
+	// TODO Check if the secret is already present etc. to bail out early
+
+	log.Info("Creating bootstrap data")
+	var (
+		files      []cloudinit.File
+		installCmd string
+	)
+
+	if config.Spec.K0s != nil {
+		//config.Spec.K0s.SetUnstructuredContent(map["spec"]interface{}{})
+		err = unstructured.SetNestedField(config.Spec.K0s.Object, scope.Cluster.Spec.ControlPlaneEndpoint.Host, "spec", "api", "externalAddress")
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error setting control plane endpoint: %v", err)
+		}
+
+		k0sConfigBytes, err := config.Spec.K0s.MarshalJSON()
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error marshalling k0s config: %v", err)
+		}
+		files = append(files, cloudinit.File{
+			Path:        "/tmp/k0s.yaml",
```

Comment:
Why on `/tmp`? These are `Machine`s and thus reboot would probably wipe out `/tmp` and render the controller bust.

Reply:
Moved it to `/etc/`